### PR TITLE
KAFKA-4455; ensure tasks are closed after CommitFailedException

### PR DESCRIPTION
**Pull request by dpoldrugo**
Details can be found in the task: [KAFKA-4455](https://issues.apache.org/jira/browse/KAFKA-4455)

Some discussion was also here: [stream shut down due to no locks for state store](https://groups.google.com/forum/#!topic/confluent-platform/i5cwYhpUtx4)

@dguy @enothereska chek it out.

### CODE REVIEW

**Comment by dpoldrugo** [comment]
```
:streams:checkstyleMain[ant:checkstyle] /home/jenkins/jenkins-slave/workspace/kafka-trunk-git-pr-jdk7/streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:392:9: '}' should be on the same line.
 FAILED
```
The code in question currenlty like this:
```java
    private void closeAllTasksOrTaskTopologies(boolean closeTopology) {
        if (closeTopology) {
            closeAllTasksTopologies();
        }
        else {
            closeAllTasks();
        }
    }
```

Should I change it to:
```java
    private void closeAllTasksOrTaskTopologies(boolean closeTopology) {
        if (closeTopology)
           closeAllTasksTopologies();
        else
            closeAllTasks();
    }
```

Or to:
```java
    private void closeAllTasksOrTaskTopologies(boolean closeTopology) {
        if (closeTopology) {
           closeAllTasksTopologies();
           return;
        }

        closeAllTasks();
    }
```

Or to:
```java
    private void closeAllTasksOrTaskTopologies(boolean closeTopology) {
        if (closeTopology) {
            closeAllTasksTopologies();
        } else {
            closeAllTasks();
        }
    }
```

**Comment by dguy** [comment]
Hi @dpoldrugo all of those options should fix the checkstyle failure, though i'd prefer the last one.

**Comment by dpoldrugo** [comment]
Imported checkstyle/checkstyle.xml - so next time this won't happen.

**Comment by mjsax** [comment]
Can you add a test?

**Comment by dpoldrugo** [comment]
Sure.
Idea for testing:
Create a Kafka Streams topology, which contains a `org.apache.kafka.streams.processor.Processor` which has a state store.
In the toplogy use a foreach action (`org.apache.kafka.streams.kstream.KStream#foreach`) which in it's `org.apache.kafka.streams.kstream.ForeachAction#apply` method calls `Thread.sleep(< ms > max.poll.interval.ms >)`

This should result in `StreamsException: stream-thread [StreamThread-1] Failed to rebalance` thrown to the `UncaughtExceptionHandler`. The cause of this exception should be `CommitFailedException: Commit cannot be completed since the group has already rebalanced` instead of `ProcessorStateException: Error opening store <storeName> at location <storePath>` with a nested cause `RocksDBException: IO error: lock <storePath>/LOCK: No locks available`

Streams config for test:
max.poll.interval.ms = \<small value\> - so the test doesn't last long, because default is 300000 (5 minutes).

What do you think of this test strategy?

**Comment by mjsax** [comment]
Sounds reasonable to me. I am hoping that the test does not introduce any test instability... We have already some issues with this. @dguy ?

**Comment by dguy** [comment]
@dpoldrugo - I wonder if we can simulate this without having to start up a full kafka streams topology? 
i.e., we have `StreamThreadTest`, which i know is a bit hard to follow, but it allows us to manually trigger the rebalance, and to manually set up the tasks. So we could have a test where we call `rebalanceListener.onPartitionsAssigned(assignedPartitions);` - we overload `StreamThread.createStreamTask` to create a task with a simple topology that just has a single state store, and we overload the `commitOffsets` method of the Task to throw the `CommitFailedException`. Keep a reference to the `StateStore` for validation
Then we invoke `rebalanceListener.onPartitionsRevoked(...)` which will cause the task to throw the exception. We then need to check that we get the `CommitFailedException` and we should also check that the store is actually closed, i.e., by calling `stateStore.isOpen(..)` and checking that it returns false or by calling `stateStore.get("key")` and checking that it throws an `InvalidStateStoreException`

**Comment by guozhangwang** [comment]
@dpoldrugo Sounds good to me. Thanks!

We are about to make a release cut for 0.10.1.1, it will be great to have this PR reviewed and merged before the deadline.

**Comment by dpoldrugo** [comment]
@dguy @mjsax I see, but could be possible that this unit test that you propose - could not cover all the aspects of this change? That's why I prefer an IT (blackbox) test more than a unit test. And one thing more, about unit testing with all those "hacks/mocks" - unit test ends up tied-up to the implementation details of the app. On the other side, and IT test, where you put an input and expect the output - you don't have to worry about internals. And later refactoring is also easier.

In the IT test , I would like to do the following:
- send 3 messages to a topic
- setup the test in a way that:
  - first message gets processed (end up in another topic)
  - second fails
  - recover the stream app and make sure that second and third message get processed in the right way
- check state store consistency with the messages in the input topic

Should I bother with this or is this scenario covered in some other IT tests?
 
@guozhangwang Hi! You mean my proposal for the testing strategy? If yes, I would like that we all agree which is the best way to go. Can you give me an ETA for the release, or just tell me how much time I have to finish this PR?

**Comment by dguy** [comment]
@dpoldrugo i'm confident all cases could be covered by unit tests.  Generally if it can be covered by unit tests then that would be the preferred method. They run much quicker, are more reliable, and are much easier to debug. It would be nice if we didn't have to worry about the internals, but the code is what it is.
Anyway, if you feel don't feel comfortable writing unit tests to cover this, then you can do it with integration tests.

**Comment by guozhangwang** [comment]
About the fix itself: my understanding is that the root cause is in `suspendTasksAndState(..)` where `rethrowException` is always set to true, whereas in `shutdownTasksAndState`, `rethrowException` is always set to false. And when `rethrowException` is always set to true, if `commitOffsets` failed then the rest of the operations are not set. So I'm wondering if a simpler fix would be just removing the `rethrowException` boolean in these two functions and always NOT rethrow exception but only log it.

About unit test: I tend to second @dguy 's proposal to do it in unit test with existing mock objects as we did in `StreamThreadTest`. The basic two cases are that we want to let the `commitOffset` throw an exception (it could be any KafkaException, not necessarily CommitFailedException) in 1) `suspendTasksAndState` and 2) `shutdownTasksAndState` functions. We do not need to check the normal `commit` call as it is already covered in other tests.

**Comment by guozhangwang** [comment]
@dpoldrugo I am trying to make a cut a release after all the "must-haves" are merged in:

https://issues.apache.org/jira/issues/?jql=project%20%3D%20KAFKA%20AND%20resolution%20%3D%20Unresolved%20AND%20fixVersion%20%3D%200.10.1.1%20ORDER%20BY%20priority%20DESC%2C%20key%20DESC

**Comment by dpoldrugo** [comment]
@dguy @guozhangwang ok, I'll write an unit test based on `StreamThreadTest`.

@guozhangwang if we set `rethrowException` to `false` or remove it, won't users be affected by this?
Then no exceptions will end-up in the user set `UncaughtExceptionHandler`, or am I wrong? And that's ok with me, as long as Kafka Streams internally knows how to recover from those exceptions. Will this be the case if we catch all `KafkaException` and don't rethrow them?

**Comment by guozhangwang** [comment]
@dpoldrugo Here is my reasoning about `rethrowException` and `UncaughtExceptionHandler`: in suspending tasks (cc @mjsax ):

1. In both `shutdownTasksAndState` and `suspendTasksAndState`, if there is any exception thrown in any of the steps, we should still proceed to try to complete the rest of the steps by all means.

2. As for whether or not we should let the user's `UncaughtExceptionHandler` be triggered, I think we should still do that in `suspendTasksAndState`, but only after all the steps are done. We could follow the example as in `KafkaProducer.close()` call, where we keep track of the `firstException` and then only throw it at the end. 

BTW we are currently working on exactly-once semantics of Streams, and under this project most of the internal exceptions should now be auto handled instead of throwing them all the way to users `UncaughtExceptionHandler`.

**Comment by dguy** [comment]
Hi @dpoldrugo are you happy to continue with this?

**Comment by dpoldrugo** [comment]
@dguy @guozhangwang @mjsax sorry for not continuing on this, I've been busy with work.
The thing is - after some time of running this code in production, I have found out that the problem still remains, as Shannon Ma pointed out in the thread: [stream shut down due to no locks for state store](https://groups.google.com/forum/#!topic/confluent-platform/i5cwYhpUtx4)

Do you have any news if this issue is still present in the master/trunk? Or it has maybe been fixed with some other pull request?

**Comment by dguy** [comment]
HI @dpoldrugo, no problem. Hopefully the problem is fixed by: https://github.com/apache/kafka/pull/2281

**Comment by guozhangwang** [comment]
Hi @dpoldrugo , could your team check if this issue still exists in current `trunk`? It would be really helpful for us to validate and continue working on it if it does.

**Comment by dpoldrugo** [comment]
@guozhangwang sure, I was thinking the same. Didn't check the code in detail, but do you think it would be possible to apply a patch to a 0.10.1.0 codebase? That would be a lot easier, since it wouldn't require us to upgrade the Kafka cluster/brokers. Or do you think the `trunk` Kafka Streams will work with Kafka cluster/brokers on `0.10.1.0`?

**Comment by guozhangwang** [comment]
My gut feeling is that 0.10.1.x broker should work with a 0.10.2.0-SNAPSHOT client since we did not make any protocol changes between the two.

**Comment by dpoldrugo** [comment]
Great. I'll try these days and get back to you.

**Comment by dpoldrugo** [comment]
Upgraded Kafka Streams to 0.10.2.0-SNAPSHOT and deployed. Your gut feeling was right, no cluster upgrade was needed. Until now - these last few hours - everithing works fine. I'll monitor how it will behave and keep you posted here.

**Comment by dpoldrugo** [comment]
> 2017-01-10 16:34:29.782  INFO 22720 --- [qtp844326063-55] o.a.kafka.common.utils.AppInfoParser     : Kafka version : 0.10.2.0-SNAPSHOT
2017-01-10 16:34:29.782  INFO 22720 --- [qtp844326063-55] o.a.kafka.common.utils.AppInfoParser     : Kafka commitId : 1fea1c390846673f

**Comment by hachikuji** [comment]
Closing this since it was resolved by #2281. Please reopen if this was a mistake.
